File: 백준/Gold/14502. 연구소/연구소.java
import java.util.*;
public class Main{
    //  0은 빈 칸, 1은 벽, 2는 바이러스
    static int N,M;
    static int[][] arr;

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();

        arr = new int[N][M];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M ; j++) {
                arr[i][j] = sc.nextInt();
            }
        }
        // dfs로 바이러스 주변에 벽을 전부 3개씩 세워보기
        dfs(0);
        // 최대값 출력
        System.out.println(max);
    }
    static int max=0;

    private static void dfs(int x) {
        if (x==3) {
            int cnt = bfs();
            max=Math.max(cnt, max);
            return;
        }
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (arr[i][j]==0) {
                    arr[i][j]=1;
                    dfs(x+1);
                    arr[i][j]=0;
                }
            }
        }
    }

    // 바이러스인 2부터 시작하여 bfs 실행
    private static int bfs() {
        Queue<int []> q = new LinkedList<>();
        boolean[][] visited = new boolean[N][M];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (arr[i][j]==2) {
                    visited[i][j]=true;
                    q.add(new int[]{i, j});
                }
            }
        }
        int[] dx = {0,0,1,-1};
        int[] dy = {1,-1,0,0};
        while(!q.isEmpty()) {
            int[] now = q.poll();
            for (int i = 0; i < 4; i++) {
                int nx = now[0]+dx[i];
                int ny = now[1]+dy[i];

                if (nx>=0 && ny>=0 && nx<N && ny<M
                && arr[nx][ny]==0 && !visited[nx][ny]) {
                    visited[nx][ny]=true;
                    q.add(new int[] {nx,ny});
                }
            }
        }
        int cnt=0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (arr[i][j]==0 && !visited[i][j]) cnt++;
            }
        }
        return cnt;
    }
}

File: 백준/Gold/14502. 연구소/연구소.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(filePath).toString().trim().split(/\s+/).map(Number);

let idx = 0;
const n = input[idx++];
const m = input[idx++];
const board = [];

for (let i = 0; i < n; i++) {
    const row = [];
    for (let j = 0; j < m; j++) {
        row.push(input[idx++]);
    }
    board.push(row);
}

let maxArea = 0;

// 벽 세우기 (조합)
function recursive(count) {
    if (count === 3) {
        maxArea = Math.max(maxArea, bfs());
        return;
    }

    for(let i=0; i<n; i++) {
        for(let j=0; j<m; j++) {
            if(board[i][j]===0) {
                board[i][j]=1;
                recursive(count+1);
                board[i][j]=0;
            }
        }
    }
}

function bfs() {
    let queue = [];
    const visited = Array.from({ length: n }, () => new Uint8Array(m));
    
    // 초기 바이러스 위치 큐에 삽입
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (board[i][j] === 2) {
                queue.push([i, j]);
                visited[i][j] = 1;
            }
        }
    }

    const dx = [0, 0, 1, -1];
    const dy = [1, -1, 0, 0];
    let head = 0;

    while (head < queue.length) {
        const [x, y] = queue[head++];

        for (let i = 0; i < 4; i++) {
            let nx = x + dx[i];
            let ny = y + dy[i];

            if (nx >= 0 && ny >= 0 && nx < n && ny < m) {
                // 원본 board는 0이고, 아직 바이러스가 방문 안 한 곳이라면
                if (board[nx][ny] === 0 && !visited[nx][ny]) {
                    visited[nx][ny] = 1;
                    queue.push([nx, ny]);
                }
            }
        }
    }

    // 안전 영역 계산: 원래 0이었던 곳 중에서 바이러스(visited)가 안 퍼진 곳
    let cnt = 0;
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            if (board[i][j] === 0 && !visited[i][j]) {
                cnt++;
            }
        }
    }
    return cnt;
}

recursive(0);
console.log(maxArea);